File: humanizer.py
import random


# Section drift profiles: how timing shifts across bars within a section
_SECTION_DRIFT = {
    "verse":       "gradual_drag",
    "atmospheric": "gradual_drag",
    "intro":       "gradual_drag",
    "quiet":       "gradual_drag",
    "outro":       "gradual_drag",
    "chorus":      "constant_push",
    "blast":       "constant_push",
    "drive":       "constant_push",
    "build":       "gradual_push",
    "buildup":     "gradual_push",
    "crescendo":   "gradual_push",
    "breakdown":   "constant_drag",
    "half_time":   "constant_drag",
    "fill":        "fill_rush",
}

# Tag → ghost clustering amount (0.0 = none, 1.0 = max)
_CLUSTER_TAG_AMOUNTS = {
    "faraquet": 0.7, "angular": 0.6, "math": 0.6,
    "raein": 0.5, "euro_screamo": 0.5, "daitro": 0.5,
    "fugazi": 0.4, "posthardcore": 0.4, "driving": 0.4,
    "screamo": 0.3, "emoviolence": 0.3,
    "shellac": 0.0, "noise_rock": 0.0, "blast": 0.0,
    "post_punk": 0.0, "motorik": 0.0,
}


def get_cluster_amount(cell):
    """Return ghost clustering amount based on cell tags. 0.0-1.0."""
    tags = cell.get("tags", [])
    amounts = [_CLUSTER_TAG_AMOUNTS[t] for t in tags if t in _CLUSTER_TAG_AMOUNTS]
    return max(amounts) if amounts else 0.3


def infer_section_type(cell):
    """Infer section type from cell tags for push/pull drift."""
    tags = cell.get("tags", [])
    if any(t in tags for t in ("blast", "extreme")):
        return "blast"
    if any(t in tags for t in ("build", "crescendo")):
        return "build"
    if any(t in tags for t in ("breakdown", "half_time")):
        return "breakdown"
    if any(t in tags for t in ("atmospheric", "sparse", "quiet")):
        return "atmospheric"
    if any(t in tags for t in ("driving", "intense")):
        return "drive"
    if any(t in tags for t in ("fill",)):
        return "fill"
    return "verse"


class Humanizer:
    VELOCITY_RANGES = {
        "ghost": (20, 50),
        "soft": (50, 75),
        "normal": (75, 105),
        "accent": (105, 127),
    }

    INSTRUMENT_VARIANCE = {
        "hihat_closed": 25,
        "hihat_open": 20,
        "ride": 22,
        "ride_bell": 20,
        "snare": 18,
        "snare_ghost": 20,
        "snare_rim": 15,
        "kick": 16,
        "crash_1": 14,
        "crash_2": 14,
        "china": 16,
        "splash": 14,
        "tom_high": 18,
        "tom_mid": 18,
        "tom_low": 18,
        "tom_floor": 18,
        "hihat_pedal": 15,
        "tom_mid_high": 18,
        "hihat_wide_open": 20,
        "crash_1_choke": 8,
        "crash_2_choke": 8,
        "ride_crash": 16,
        "fx_cymbal_1": 14,
        "fx_cymbal_2": 14,
        "china_2": 16,
    }

    TIMING_TENDENCIES = {
        "snare": 2,
        "snare_ghost": 4,
        "snare_rim": 2,
        "kick": 0,
        "hihat_closed": -1,
        "ride": -2,
        "ride_bell": -2,
        "crash_1": -3,
        "crash_2": -3,
        "china": -2,
        "tom_floor": 1,
        "tom_low": 1,
        "tom_mid": 0,
        "tom_high": 0,
        "hihat_pedal": 0,
        "hihat_open": -1,
        "splash": -2,
        "tom_mid_high": 0,
        "hihat_wide_open": -1,
        "crash_1_choke": 0,
        "crash_2_choke": 0,
        "ride_crash": -3,
        "fx_cymbal_1": -2,
        "fx_cymbal_2": -2,
        "china_2": -2,
    }

    CONTOUR_INSTRUMENTS = {"ride", "ride_bell", "hihat_closed", "hihat_open", "hihat_wide_open"}

    CONTOUR_OFFSETS = {
        0.0:  +6,   # downbeat — strongest
        0.25: -4,   # first sixteenth — weak
        0.5:  +3,   # eighth — medium strong
        0.75: -5,   # third sixteenth — weakest
    }

    def __init__(self, humanize_amount, seed=None):
        self.rng = random.Random(seed)
        self.humanize_amount = humanize_amount

    def humanize_velocity(self, velocity_level, instrument):
        level = velocity_level.lower()
        if level not in self.VELOCITY_RANGES:
            level = "normal"
        low, high = self.VELOCITY_RANGES[level]
        center = (low + high) // 2
        variance = self.INSTRUMENT_VARIANCE.get(instrument, 12)
        scaled_variance = max(3, int(variance * self.humanize_amount))
        if scaled_variance <= 0:
            return max(1, min(127, center))
        vel = self.rng.randint(center - scaled_variance, center + scaled_variance)
        return max(1, min(127, vel))

    def humanize_timing(self, abs_tick, instrument, tempo_bpm, ppq=480):
        ms_per_tick = (60000.0 / tempo_bpm) / ppq
        tendency_ms = self.TIMING_TENDENCIES.get(instrument, 0)
        tendency_ticks = int(tendency_ms * self.humanize_amount / ms_per_tick)
        jitter_ms = self.rng.gauss(0, 6 * self.humanize_amount)
        jitter_ticks = int(jitter_ms / ms_per_tick)
        return max(0, abs_tick + tendency_ticks + jitter_ticks)

    def apply_swing(self, abs_tick, is_upbeat, swing_amount, ticks_per_beat):
        if not is_upbeat or swing_amount <= 0:
            return abs_tick
        triplet_offset = ticks_per_beat // 3
        swing_offset = int(triplet_offset * swing_amount)
        return abs_tick + swing_offset

    def velocity_contour(self, velocity, instrument, beat, sub):
        """Apply wrist-pattern contour + beat-1 emphasis to cymbal instruments."""
        if instrument not in self.CONTOUR_INSTRUMENTS:
            return velocity
        contour_amount = self.humanize_amount * 0.8
        if contour_amount == 0:
            return velocity
        closest_sub = min(self.CONTOUR_OFFSETS, key=lambda s: abs(s - sub))
        offset = self.CONTOUR_OFFSETS[closest_sub] * contour_amount
        if beat == 1:
            offset += 4 * contour_amount
        return max(1, min(127, velocity + int(offset)))

    def compute_section_drift_ms(self, section_type, bar_index, total_bars):
        """Compute timing drift in ms for a bar based on section type."""
        drift_amount = self.humanize_amount * 0.7
        if drift_amount == 0:
            return 0.0
        profile = _SECTION_DRIFT.get(section_type)
        if not profile:
            return 0.0
        progress = bar_index / max(1, total_bars - 1) if total_bars > 1 else 0.0
        if profile == "gradual_drag":
            return min(6.0, bar_index * 1.0) * drift_amount
        elif profile == "constant_push":
            return -4.0 * drift_amount
        elif profile == "constant_drag":
            return 5.0 * drift_amount
        elif profile == "gradual_push":
            return -6.0 * progress * drift_amount
        elif profile == "fill_rush":
            return -8.0 * progress * drift_amount
        return 0.0

    def apply_flam(self, events, tempo, ppq):
        """Pull kick earlier when simultaneous with snare. Returns new list."""
        if self.humanize_amount < 0.2:
            return list(events)
        ms_per_tick = (60000.0 / tempo) / ppq
        tick_map = {}
        for i, (tick, inst, vel) in enumerate(events):
            tick_map.setdefault(tick, []).append(i)
        result = list(events)
        for tick, indices in tick_map.items():
            instruments = {result[i][1]: i for i in indices}
            velocities = {result[i][1]: result[i][2] for i in indices}
            has_kick = "kick" in instruments and velocities.get("kick", 0) > 75
            has_snare = any(
                inst in instruments and velocities.get(inst, 0) > 75
                for inst in ("snare", "snare_rim")
            )
            if has_kick and has_snare:
                flam_ms = self.rng.uniform(5, 12)
                flam_ticks = max(1, int(flam_ms / ms_per_tick))
                ki = instruments["kick"]
                old = result[ki]
                result[ki] = (max(0, old[0] - flam_ticks), old[1], old[2])
        return result

    def apply_ghost_clustering(self, events, cluster_amount, tempo, ppq):
        """Pull ghost notes toward nearby snare accents. Returns new list."""
        effective = cluster_amount * self.humanize_amount
        if effective <= 0:
            return list(events)
        sixteenth_ticks = ppq // 4
        ghosts, accents, others = [], [], []
        for i, (tick, inst, vel) in enumerate(events):
            if inst == "snare_ghost" or (inst == "snare" and vel < 50):
                ghosts.append([tick, inst, vel])
            elif inst in ("snare", "snare_rim") and vel >= 100:
                accents.append((tick, inst, vel))
                others.append((tick, inst, vel))
            else:
                others.append((tick, inst, vel))
        if not accents or not ghosts:
            return list(events)
        accent_ticks = sorted(set(t for t, _, _ in accents))
        new_ghosts = []
        all_ticks = set(e[0] for e in events)
        for ghost in ghosts:
            nearest = min(accent_ticks, key=lambda t: abs(t - ghost[0]))
            distance = nearest - ghost[0]
            pull = int(distance * effective * 0.4)
            pull = max(-sixteenth_ticks, min(sixteenth_ticks, pull))
            ghost[0] = max(0, ghost[0] + pull)
            ghost[0] = max(0, ghost[0] + self.rng.randint(-5, 5))
            if self.rng.random() < effective * 0.3:
                if distance > 0:
                    new_tick = nearest + self.rng.randint(sixteenth_ticks // 2, sixteenth_ticks)
                else:
                    new_tick = nearest - self.rng.randint(sixteenth_ticks // 2, sixteenth_ticks)
                new_tick = max(0, new_tick)
                if not any(abs(new_tick - t) < sixteenth_ticks // 4 for t in all_ticks):
                    new_vel = self.rng.randint(20, 45)
                    new_ghosts.append((new_tick, "snare_ghost", new_vel))
                    all_ticks.add(new_tick)
        return others + [tuple(g) for g in ghosts] + new_ghosts

File: plugin/src/engine/humanizer.rs
use rand::Rng;
use rand::distributions::{Distribution, Standard};
use rand_chacha::ChaCha8Rng;
use rand::SeedableRng;

use super::cell::{Instrument, VelocityLevel};

/// Section drift profiles — how timing shifts across bars within a section.
fn drift_profile(section_type: &str) -> Option<&'static str> {
    match section_type {
        "verse" | "atmospheric" | "intro" | "quiet" | "outro" => Some("gradual_drag"),
        "chorus" | "blast" | "drive" => Some("constant_push"),
        "build" | "buildup" | "crescendo" => Some("gradual_push"),
        "breakdown" | "half_time" => Some("constant_drag"),
        "fill" => Some("fill_rush"),
        _ => None,
    }
}

/// Tag → ghost clustering amount (0.0 = none, 1.0 = max).
fn cluster_tag_amount(tag: &str) -> Option<f64> {
    match tag {
        "faraquet" => Some(0.7),
        "angular" => Some(0.6),
        "math" => Some(0.6),
        "raein" | "euro_screamo" | "daitro" => Some(0.5),
        "fugazi" | "posthardcore" | "driving" => Some(0.4),
        "screamo" | "emoviolence" => Some(0.3),
        "shellac" | "noise_rock" | "blast" | "post_punk" | "motorik" => Some(0.0),
        _ => None,
    }
}

/// Get ghost clustering amount for a cell based on its tags.
pub fn get_cluster_amount(tags: &[String]) -> f64 {
    let amounts: Vec<f64> = tags.iter()
        .filter_map(|t| cluster_tag_amount(t))
        .collect();
    if amounts.is_empty() {
        0.3
    } else {
        amounts.iter().cloned().fold(f64::NEG_INFINITY, f64::max)
    }
}

/// Infer section type from cell tags for push/pull drift.
pub fn infer_section_type(tags: &[String]) -> &'static str {
    let has = |t: &str| tags.iter().any(|tag| tag == t);
    if has("blast") || has("extreme") { return "blast"; }
    if has("build") || has("crescendo") { return "build"; }
    if has("breakdown") || has("half_time") { return "breakdown"; }
    if has("atmospheric") || has("sparse") || has("quiet") { return "atmospheric"; }
    if has("driving") || has("intense") { return "drive"; }
    if has("fill") { return "fill"; }
    "verse"
}

/// An assembled MIDI event: (absolute_tick, instrument, velocity).
#[derive(Debug, Clone)]
pub struct Event {
    pub tick: i64,
    pub instrument: Instrument,
    pub velocity: i32,
}

/// Physics-based drum humanizer with seeded RNG.
pub struct Humanizer {
    pub rng: ChaCha8Rng,
    pub humanize_amount: f64,
}

impl Humanizer {
    pub fn new(humanize_amount: f64, seed: u64) -> Self {
        Self {
            rng: ChaCha8Rng::seed_from_u64(seed),
            humanize_amount,
        }
    }

    /// Per-instrument velocity variance width.
    fn instrument_variance(instrument: Instrument) -> i32 {
        match instrument {
            Instrument::HihatClosed => 25,
            Instrument::HihatOpen => 20,
            Instrument::Ride => 22,
            Instrument::RideBell => 20,
            Instrument::Snare => 18,
            Instrument::SnareGhost => 20,
            Instrument::SnareRim => 15,
            Instrument::Kick => 16,
            Instrument::Crash1 | Instrument::Crash2 => 14,
            Instrument::China => 16,
            Instrument::Splash => 14,
            Instrument::TomHigh | Instrument::TomMid | Instrument::TomLow | Instrument::TomFloor => 18,
            Instrument::TomMidHigh => 18,
            Instrument::HihatPedal => 15,
            Instrument::HihatWideOpen => 20,
            Instrument::Crash1Choke | Instrument::Crash2Choke => 8,
            Instrument::RideCrash => 16,
            Instrument::FxCymbal1 | Instrument::FxCymbal2 => 14,
            Instrument::China2 => 16,
        }
    }

    /// Per-instrument timing tendency in milliseconds.
    fn timing_tendency(instrument: Instrument) -> f64 {
        match instrument {
            Instrument::Snare => 2.0,
            Instrument::SnareGhost => 4.0,
            Instrument::SnareRim => 2.0,
            Instrument::Kick => 0.0,
            Instrument::HihatClosed => -1.0,
            Instrument::Ride => -2.0,
            Instrument::RideBell => -2.0,
            Instrument::Crash1 | Instrument::Crash2 => -3.0,
            Instrument::China => -2.0,
            Instrument::TomFloor | Instrument::TomLow => 1.0,
            Instrument::TomMid | Instrument::TomHigh | Instrument::TomMidHigh => 0.0,
            Instrument::HihatPedal => 0.0,
            Instrument::HihatOpen => -1.0,
            Instrument::Splash => -2.0,
            Instrument::HihatWideOpen => -1.0,
            Instrument::Crash1Choke | Instrument::Crash2Choke => 0.0,
            Instrument::RideCrash => -3.0,
            Instrument::FxCymbal1 | Instrument::FxCymbal2 => -2.0,
            Instrument::China2 => -2.0,
        }
    }

    /// Contour offset for wrist pattern on cymbals.
    fn contour_offset(sub: f64) -> f64 {
        // Find closest sub value
        let subs = [(0.0, 6.0), (0.25, -4.0), (0.5, 3.0), (0.75, -5.0)];
        let mut closest = subs[0];
        let mut min_dist = f64::MAX;
        for &(s, offset) in &subs {
            let dist = (s - sub).abs();
            if dist < min_dist {
                min_dist = dist;
                closest = (s, offset);
            }
        }
        closest.1
    }

    /// Whether an instrument gets velocity contour (wrist pattern).
    fn is_contour_instrument(instrument: Instrument) -> bool {
        matches!(
            instrument,
            Instrument::Ride | Instrument::RideBell
                | Instrument::HihatClosed | Instrument::HihatOpen
                | Instrument::HihatWideOpen
        )
    }

    /// Generate a random integer in [low, high] inclusive.
    fn randint(&mut self, low: i32, high: i32) -> i32 {
        if low >= high { return low; }
        self.rng.gen_range(low..=high)
    }

    /// Generate a gaussian random value with given mean and stddev.
    fn gauss(&mut self, mean: f64, stddev: f64) -> f64 {
        use rand_distr::Normal;
        let normal = Normal::new(mean, stddev).unwrap_or(Normal::new(0.0, 1.0).unwrap());
        normal.sample(&mut self.rng)
    }

    /// Generate a uniform random f64 in [low, high].
    fn uniform(&mut self, low: f64, high: f64) -> f64 {
        self.rng.gen_range(low..=high)
    }

    /// Generate a random f64 in [0.0, 1.0).
    fn random(&mut self) -> f64 {
        self.rng.gen::<f64>()
    }

    /// Humanize velocity for a given level and instrument.
    /// Returns a MIDI velocity value (1-127).
    pub fn humanize_velocity(&mut self, velocity_level: VelocityLevel, instrument: Instrument) -> i32 {
        let (low, high) = velocity_level.range();
        let center = (low + high) / 2;
        let variance = Self::instrument_variance(instrument);
        let scaled_variance = (variance as f64 * self.humanize_amount).round() as i32;
        let scaled_variance = scaled_variance.max(3);
        if scaled_variance <= 0 {
            return center.clamp(1, 127);
        }
        let vel = self.randint(center - scaled_variance, center + scaled_variance);
        vel.clamp(1, 127)
    }

    /// Humanize timing for a hit at the given absolute tick position.
    /// Returns adjusted absolute tick (clamped to >= 0).
    pub fn humanize_timing(&mut self, abs_tick: i64, instrument: Instrument, tempo_bpm: f64, ppq: i64) -> i64 {
        let ms_per_tick = (60000.0 / tempo_bpm) / ppq as f64;
        let tendency_ms = Self::timing_tendency(instrument);
        let tendency_ticks = (tendency_ms * self.humanize_amount / ms_per_tick) as i64;
        let jitter_ms = self.gauss(0.0, 6.0 * self.humanize_amount);
        let jitter_ticks = (jitter_ms / ms_per_tick) as i64;
        (abs_tick + tendency_ticks + jitter_ticks).max(0)
    }

    /// Apply swing to upbeat hits.
    pub fn apply_swing(&self, abs_tick: i64, is_upbeat: bool, swing_amount: f64, ticks_per_beat: i64) -> i64 {
        if !is_upbeat || swing_amount <= 0.0 {
            return abs_tick;
        }
        let triplet_offset = ticks_per_beat / 3;
        let swing_offset = (triplet_offset as f64 * swing_amount) as i64;
        abs_tick + swing_offset
    }

    /// Apply wrist-pattern contour + beat-1 emphasis to cymbal instruments.
    pub fn velocity_contour(&self, velocity: i32, instrument: Instrument, beat: i32, sub: f64) -> i32 {
        if !Self::is_contour_instrument(instrument) {
            return velocity;
        }
        let contour_amount = self.humanize_amount * 0.8;
        if contour_amount == 0.0 {
            return velocity;
        }
        let mut offset = Self::contour_offset(sub) * contour_amount;
        if beat == 1 {
            offset += 4.0 * contour_amount;
        }
        (velocity + offset as i32).clamp(1, 127)
    }

    /// Compute timing drift in ms for a bar based on section type.
    pub fn compute_section_drift_ms(&self, section_type: &str, bar_index: i32, total_bars: i32) -> f64 {
        let drift_amount = self.humanize_amount * 0.7;
        if drift_amount == 0.0 {
            return 0.0;
        }
        let profile = match drift_profile(section_type) {
            Some(p) => p,
            None => return 0.0,
        };
        let progress = if total_bars > 1 {
            bar_index as f64 / (total_bars - 1) as f64
        } else {
            0.0
        };
        match profile {
            "gradual_drag" => (bar_index as f64 * 1.0).min(6.0) * drift_amount,
            "constant_push" => -4.0 * drift_amount,
            "constant_drag" => 5.0 * drift_amount,
            "gradual_push" => -6.0 * progress * drift_amount,
            "fill_rush" => -8.0 * progress * drift_amount,
            _ => 0.0,
        }
    }

    /// Pull kick earlier when simultaneous with high-velocity snare. Returns new event list.
    pub fn apply_flam(&mut self, events: &[Event], tempo: f64, ppq: i64) -> Vec<Event> {
        if self.humanize_amount < 0.2 {
            return events.to_vec();
        }
        let ms_per_tick = (60000.0 / tempo) / ppq as f64;

        // Group events by tick
        let mut tick_map: std::collections::HashMap<i64, Vec<usize>> = std::collections::HashMap::new();
        for (i, event) in events.iter().enumerate() {
            tick_map.entry(event.tick).or_default().push(i);
        }

        let mut result = events.to_vec();

        for (_tick, indices) in &tick_map {
            let mut instruments: std::collections::HashMap<Instrument, usize> = std::collections::HashMap::new();
            let mut velocities: std::collections::HashMap<Instrument, i32> = std::collections::HashMap::new();

            for &i in indices {
                instruments.insert(result[i].instrument, i);
                velocities.insert(result[i].instrument, result[i].velocity);
            }

            let has_kick = instruments.contains_key(&Instrument::Kick)
                && *velocities.get(&Instrument::Kick).unwrap_or(&0) > 75;
            let has_snare = [Instrument::Snare, Instrument::SnareRim].iter().any(|inst| {
                instruments.contains_key(inst) && *velocities.get(inst).unwrap_or(&0) > 75
            });

            if has_kick && has_snare {
                let flam_ms = self.uniform(5.0, 12.0);
                let flam_ticks = (flam_ms / ms_per_tick) as i64;
                let flam_ticks = flam_ticks.max(1);
                if let Some(&ki) = instruments.get(&Instrument::Kick) {
                    result[ki].tick = (result[ki].tick - flam_ticks).max(0);
                }
            }
        }

        result
    }

    /// Pull ghost notes toward nearby snare accents. Returns new event list.
    pub fn apply_ghost_clustering(&mut self, events: &[Event], cluster_amount: f64, _tempo: f64, ppq: i64) -> Vec<Event> {
        let effective = cluster_amount * self.humanize_amount;
        if effective <= 0.0 {
            return events.to_vec();
        }
        let sixteenth_ticks = ppq / 4;

        let mut ghosts: Vec<Event> = Vec::new();
        let mut accents: Vec<Event> = Vec::new();
        let mut others: Vec<Event> = Vec::new();

        for event in events {
            if event.instrument == Instrument::SnareGhost
                || (event.instrument == Instrument::Snare && event.velocity < 50)
            {
                ghosts.push(event.clone());
            } else if (event.instrument == Instrument::Snare || event.instrument == Instrument::SnareRim)
                && event.velocity >= 100
            {
                accents.push(event.clone());
                others.push(event.clone());
            } else {
                others.push(event.clone());
            }
        }

        if accents.is_empty() || ghosts.is_empty() {
            return events.to_vec();
        }

        let mut accent_ticks: Vec<i64> = accents.iter().map(|e| e.tick).collect();
        accent_ticks.sort();
        accent_ticks.dedup();

        let mut new_ghosts: Vec<Event> = Vec::new();
        let mut all_ticks: std::collections::HashSet<i64> = events.iter().map(|e| e.tick).collect();

        for ghost in &mut ghosts {
            let nearest = *accent_ticks.iter()
                .min_by_key(|&&t| (t - ghost.tick).abs())
                .unwrap();
            let distance = nearest - ghost.tick;
            let pull = ((distance as f64 * effective * 0.4) as i64)
                .clamp(-sixteenth_ticks, sixteenth_ticks);
            ghost.tick = (ghost.tick + pull).max(0);
            let jitter = self.randint(-5, 5) as i64;
            ghost.tick = (ghost.tick + jitter).max(0);

            if self.random() < effective * 0.3 {
                let new_tick = if distance > 0 {
                    nearest + self.randint(sixteenth_ticks as i32 / 2, sixteenth_ticks as i32) as i64
                } else {
                    nearest - self.randint(sixteenth_ticks as i32 / 2, sixteenth_ticks as i32) as i64
                };
                let new_tick = new_tick.max(0);
                let too_close = all_ticks.iter().any(|&t| (new_tick - t).abs() < sixteenth_ticks / 4);
                if !too_close {
                    let new_vel = self.randint(20, 45);
                    new_ghosts.push(Event {
                        tick: new_tick,
                        instrument: Instrument::SnareGhost,
                        velocity: new_vel,
                    });
                    all_ticks.insert(new_tick);
                }
            }
        }

        let mut result = others;
        result.extend(ghosts);
        result.extend(new_ghosts);
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_velocity_ranges() {
        let mut h = Humanizer::new(0.5, 42);
        // Accent should be in the high range
        for _ in 0..20 {
            let vel = h.humanize_velocity(VelocityLevel::Accent, Instrument::Snare);
            assert!(vel >= 1 && vel <= 127);
        }
        // Ghost should be in the low range
        for _ in 0..20 {
            let vel = h.humanize_velocity(VelocityLevel::Ghost, Instrument::Snare);
            assert!(vel >= 1 && vel <= 127);
        }
    }

    #[test]
    fn test_humanize_at_zero() {
        let mut h = Humanizer::new(0.0, 42);
        // At humanize=0, timing should still work but with minimal jitter
        let tick = h.humanize_timing(1000, Instrument::Kick, 120.0, 480);
        // Should be close to 1000 (kick has 0ms tendency)
        assert!((tick - 1000).abs() < 50);
    }

    #[test]
    fn test_velocity_contour() {
        let h = Humanizer::new(0.5, 42);
        let base_vel = 90;
        // Beat 1, sub 0.0 should get a boost (contour +6 * 0.4 + beat1 +4 * 0.4)
        let boosted = h.velocity_contour(base_vel, Instrument::Ride, 1, 0.0);
        assert!(boosted > base_vel);
        // Non-cymbal should be unchanged
        let unchanged = h.velocity_contour(base_vel, Instrument::Kick, 1, 0.0);
        assert_eq!(unchanged, base_vel);
    }

    #[test]
    fn test_section_drift() {
        let h = Humanizer::new(0.5, 42);
        // Verse should drag (positive ms)
        let drift = h.compute_section_drift_ms("verse", 3, 8);
        assert!(drift > 0.0);
        // Blast should push (negative ms)
        let drift = h.compute_section_drift_ms("blast", 3, 8);
        assert!(drift < 0.0);
    }

    #[test]
    fn test_infer_section_type() {
        assert_eq!(infer_section_type(&["blast".into(), "intense".into()]), "blast");
        assert_eq!(infer_section_type(&["driving".into()]), "drive");
        assert_eq!(infer_section_type(&["groovy".into()]), "verse");
    }

    #[test]
    fn test_cluster_amount() {
        assert!((get_cluster_amount(&["faraquet".into()]) - 0.7).abs() < 0.001);
        assert!((get_cluster_amount(&["shellac".into()]) - 0.0).abs() < 0.001);
        assert!((get_cluster_amount(&["unknown_tag".into()]) - 0.3).abs() < 0.001);
    }
}
